crates/node-core/src/init.rs

```diff
@@ -379,10 +379,53 @@ pub fn init_from_state_dump<DB: Database>(
         line.clear();
     }
 
+    // compute and compare state root
+    //
+    // note: we compute it even if we have nothing to compare to because we set the stage
+    // checkpoints further up.
+    let computed_state_root = compute_state_root(&provider_rw)?;
+    if let Some(expected_state_root) = expected_state_root {
+        if computed_state_root != expected_state_root {
+            error!(target: "reth::cli", ?computed_state_root, ?expected_state_root, "Computed state root does not match state root in state dump");
+            eyre::bail!("Computed state root differs from state root in state dump. Got {:?}, expected {:?}", computed_state_root, expected_state_root);
+        }
+    } else {
+        warn!(target: "reth::cli", "No state root found in file, skipping verification.");
+    }
+    provider_rw.commit()?;
+
     Ok(hash)
 }
 
-#[derive(Debug, Serialize, Deserialize)]
+/// Computes the state root (from scratch) based on the accounts and storages present in the
+/// database.
+fn compute_state_root<DB: Database>(provider: &DatabaseProviderRW<DB>) -> eyre::Result<StateRoot> {
+    trace!(target: "reth::cli", "Computing state root");
+
+    let tx = provider.tx_ref();
+    let mut intermediate_state: Option<IntermediateStateRootState> = None;
+    let root = loop {
+        match StateRootComputer::from_tx(tx)
+            .with_intermediate_state(intermediate_state)
+            .root_with_progress()?
+        {
+            StateRootProgress::Progress(state, _, updates) => {
+                trace!(target: "reth::cli", "Flushing trie updates");
```

Comment:
```suggestion
                trace!(target: "reth::cli", last_account_key = %state.last_account_key, updates_len = updates.len(), "Flushing trie updates");
```

Reply:
@onbjerg last_account_key would give you visibility into the progress of the computation